satpy/modifiers/angles.py

```diff
@@ -0,0 +1,266 @@
+    for arg in args:
+        if isinstance(arg, datetime):
+            new_args.append(STATIC_EARTH_INERTIAL_DATETIME)
+        elif isinstance(arg, (float, np.float64, np.float32)):
+            # round floating point numbers to nearest tenth
+            new_args.append(round(arg, 1))
+        else:
+            new_args.append(arg)
+    return new_args
+
+
+def _geo_dask_to_data_array(arr: da.Array) -> xr.DataArray:
+    return xr.DataArray(arr, dims=('y', 'x'))
+
+
+def get_angles(data_arr: xr.DataArray) -> tuple[xr.DataArray, xr.DataArray, xr.DataArray, xr.DataArray]:
+    """Get sun and satellite angles to use in crefl calculations.
+
+    Note that this function can benefit from the ``satpy.config`` parameters
+    :ref:`cache_lonlats <config_cache_lonlats_setting>` and
+    :ref:`cache_sensor_angles <config_cache_sensor_angles_setting>`
+    being set to ``True``.
+
+    """
+    sun_angles = _get_sun_angles(data_arr)
+    sat_angles = _get_sensor_angles(data_arr)
+    # sata, satz, suna, sunz
+    return sat_angles + sun_angles
+
+
+def get_satellite_zenith_angle(data_arr: xr.DataArray) -> xr.DataArray:
+    """Generate satellite zenith angle for the provided data.
+
+    Note that this function can benefit from the ``satpy.config`` parameters
+    :ref:`cache_lonlats <config_cache_lonlats_setting>` and
+    :ref:`cache_sensor_angles <config_cache_sensor_angles_setting>`
+    being set to ``True``.
+
+    """
+    satz = _get_sensor_angles(data_arr)[1]
+    return satz
+
+
+@cache_to_zarr("cache_lonlats")
+def _get_valid_lonlats(area: PRGeometry, chunks: int = "auto") -> tuple[da.Array, da.Array]:
+    with ignore_invalid_float_warnings():
+        lons, lats = area.get_lonlats(chunks=chunks)
+        lons = da.where(lons >= 1e30, np.nan, lons)
+        lats = da.where(lats >= 1e30, np.nan, lats)
+    return lons, lats
+
+
+def _get_sun_angles(data_arr: xr.DataArray) -> tuple[xr.DataArray, xr.DataArray]:
+    lons, lats = _get_valid_lonlats(data_arr.attrs["area"], data_arr.data.chunks)
+    with ignore_invalid_float_warnings():
+        suna = get_alt_az(data_arr.attrs['start_time'], lons, lats)[1]
+        suna = np.rad2deg(suna)
+        sunz = sun_zenith_angle(data_arr.attrs['start_time'], lons, lats)
+    suna = _geo_dask_to_data_array(suna)
+    sunz = _geo_dask_to_data_array(sunz)
+    return suna, sunz
+
+
+def _get_sensor_angles(data_arr: xr.DataArray) -> tuple[xr.DataArray, xr.DataArray]:
+    sat_lon, sat_lat, sat_alt = get_satpos(data_arr)
+    area_def = data_arr.attrs["area"]
+    sata, satz = _get_sensor_angles_from_sat_pos(sat_lon, sat_lat, sat_alt,
+                                                 data_arr.attrs["start_time"],
+                                                 area_def, data_arr.data.chunks)
+    sata = _geo_dask_to_data_array(sata)
+    satz = _geo_dask_to_data_array(satz)
+    return sata, satz
+
+
+@cache_to_zarr("cache_sensor_angles", sanitize_args_func=_sanitize_observer_look_args)
```

Comment:
I'd rather have two decorators here, one for argument rounding, and one for the caching if possible.

Reply:
Agreed.

Reply:
I don't think I can do that. We only want to round the arguments if we are caching. We can't know if we are caching unless we're in the caching logic and the caching logic can't know if it can cache unless it has the sanitized arguments (dealing with datetimes or DataArray inputs) and knows the `cache_dir` to use.

Reply:
the way I thought about is it would look like:
```
@sanitize_oberver_look_args_if("cache_sensor_angles")
@cache_to_zarr_if("cache_sensor_angles")
...
```

Reply:
I like the idea, but I think it may end up being a lot of duplicated code. To purely sanitize the arguments based on this config value would be pretty easy. There is the additional complexity of `cache_dir` needing to be defined, but I guess if I assume that `cache_dir` (from `satpy.config`) is always defined then that removes that chunk of code. The only other complication then is the "uncacheable" types. The sanitizing decorator should only return sanitized arguments if the config parameter is `True` and none of the arguments are of the "uncacheable" types. This means that both the cache_to_zarr_if decorator and the sanitize decorator would have to know these types. If you're OK with these workarounds then I'm ok implementing it @mraspaud. Let me know.

Other thing to keep in mind is that this would mean that the order of the decorators matters a lot.

Reply:
If it adds a lot of complexity I don't think it's worth it. So you can leave it be.

Reply:
I think it does add complexity, duplicate code, and misdirection (I'm biased as the person who wrote the original method). So I think I'm going to leave it. I take this as an approval from both of you then so this can be merged. I'll finish up another project and then merge this later in case you have other concerns.